### DISCUSSION TITLE
Next middleware makes countless redirect requests

### DISCUSSION THREAD

**Discussion by SEOKKAMONI** [vercel/next.js#54591] [Help] [answered]
### Summary

![image](https://github.com/vercel/next.js/assets/102217654/15ecb0d3-b665-41dc-840e-ca4e7ca2ae96)

EOM

### Additional information

```js
import type { NextRequest } from 'next/server';
import { NextResponse, userAgent } from 'next/server';

export const middleware = (request: NextRequest) => {
    const url = request.url;
    const { device } = userAgent(request);

    if (device.type === 'mobile' && url !== '/mobile') {
        return NextResponse.redirect(new URL('/mobile', request.url));
    }
};
```

### Example

https://github.com/Bamdoliro/marururu/blob/develop/apps/user/src/middleware.ts

https://github.com/Bamdoliro/marururu/blob/develop/apps/user/src/app/mobile/page.tsx

_No response_

**Comment by hyjoong** [comment, marked as answer]
If you haven't figure out matcher for your middleware, it will be triggered for every route in your project. To control which middleware should or should not run for specific routes, you should set the matcher in the middleware's configuration.

Also, the code snippet if (device.type === 'mobile' && url !== '/mobile') seems to be incorrectly written. This is because, if you log request.url, you will notice that it represents the entire URL path. For instance, the URL might be http://localhost:3000/mobile. Therefore, it seems more appropriate to use request.nextUrl.pathname instead of url. I believe that configuring the matcher and using request.nextUrl.pathname should resolve the issue.


https://nextjs.org/docs/app/building-your-application/routing/middleware#matcher

example code
```javascript
export const middleware = (request: NextRequest) => {
  const url = request.url;
  const { device } = userAgent(request);

 if (device.type === 'mobile' && request.nextUrl.pathname !== '/mobile') {
      return NextResponse.redirect(new URL('/mobile', request.url));
  }
}

export const config = {
  matcher: ['/mobile', '/login', 'result/:path*'],
}

// or

export const config = {
  matcher: ['/((?!api|_next/static|_next/image|favicon.ico).*)']
}

```

```diff
-  if (device.type === 'mobile' && url !== '/mobile') {
+ if (device.type === 'mobile' && request.nextUrl.pathname !== '/mobile') {

+ export const config = {
+  matcher: ['/((?!api|_next/static|_next/image|favicon.ico).*)']
+ }
```

**Reply by hyjoong** [reply]
The matcher code written above is just an example, and depending on the project's requirements, you may need to add more paths that should not be executed. Upon reviewing the code in the shared repository, it appears that SVG files are also being redirected, which could potentially cause issues. You can add SVG files to the exclusion list as shown in the example below:

```javascript
// add svg
export const config = {
    matcher: ['/((?!api|_next/static|_next/image|favicon.ico|svg).*)'],
};
```

**Reply by SEOKKAMONI** [reply]
I solved it thank you for hyjoong!!
